### ISSUE TITLE
How to use dictionary without ID?

### ISSUE THREAD

**Issue by pmeenan** [facebook/zstd#3630]
We're in the process of working on a [HTTP standard for negotiating dictionary-based content-encoding](https://github.com/WICG/compression-dictionary-transport) with the expectation that it should be able to support any compression algorithm that supports external dictionaries (Brotli and Zstandard initially).

The dictionaries that are negotiated support both [purpose-built dictionaries](https://github.com/WICG/compression-dictionary-transport/blob/main/README.md#dynamic-resources-flow) as well as using a [previously-cached response as a dictionary for future requests](https://github.com/WICG/compression-dictionary-transport/blob/main/README.md#static-resources-flow) (i.e. a new version of a Javascript bundle using the previous version as a dictionary).

For the purpose-built dictionary case, I could see private dictionary ID's working if the negotiated dictionaries are pre-prepared and not just a text file but is there a way to specify dictionaries that are identified out-of-band (we use sha-256 hashes for example) so that stand-alone text files can be used as a dictionary on both sides of the stream?

Should we reserve a specific ID to be used for both ends of this?

Our initial trials are likely to use Brotli since Chrome already ships with the necessary bits but I'd like to make sure the spec proposal handles the Zstandard case as well.

**Comment by felixhandte** [comment]
TL;DR: I believe the proposal as written will work with Zstd.

Compression Dictionary Transport, as I understand it, does not rely on any signalling mechanism in the compressed stream. Whether a dictionary was used, and which, are all negotiated explicitly at the HTTP layer. So as far as I can tell nothing is required of the compressor other than to be a passive recipient of the dictionary and content blobs and do the correct compression and decompression operation on them.

The only remaining question would be whether zstd's internal dictionary signalling/coordination mechanisms are in conflict with this external negotiation. On that front, I believe there shouldn't be any issues:

Zstandard supports two kinds of dictionaries, raw dictionaries and formatted dictionaries. Raw dictionaries do not have an ID from Zstandard's perspective and no ID is written into [the frame header](https://www.rfc-editor.org/rfc/rfc8878#name-dictionary_id). Formatted dictionaries do have [a 32-bit int ID](https://www.rfc-editor.org/rfc/rfc8878#section-5-6.4.1), and unless [instructed otherwise](https://github.com/facebook/zstd/blob/v1.5.5/lib/zstd.h#L429) this ID will be written into the frame header. At decode-time, if a dictionary ID is present in the frame header, zstd will require that a dictionary with the same ID was presented to the decoder. As long as the same bytestream is presented to the encoder and decoder, CDT should never run afoul of these restrictions.

Distinguishing between a raw dictionary and a formatted dictionary is likely something we'll want to do (likely paralleling the purpose-built vs re-use) usecases. But I think it doesn't need to happen in the CDT proposal. Instead, in spec'ing a use of CDT for Zstandard, we might add a behavior where "unless the response being used as a dictionary has a MIME type of `application/zstd-dict` (although no such type is currently defined...), the dictionary content must be presented to the compressor as a 'raw dictionary' (i.e., [`ZSTD_dct_rawContent`](https://github.com/facebook/zstd/blob/v1.5.5/lib/zstd.h#L1312))."

**Comment by pmeenan** [comment]
Awesome, thanks.
